opte/src/engine/packet.rs

```diff
@@ -2622,4 +2620,48 @@ mod test {
         assert_eq!(pkt.num_segs(), 2);
         assert!(matches!(pkt.parse(), Err(ParseError::BadHeader(_))));
     }
+
+    // Verify that we correctly parse an IPv6 packet with extension headers
+    #[test]
+    fn parse_ipv6_extension_headers_ok() {
+        use crate::engine::ip6::test::generate_test_packet;
+        use crate::engine::ip6::test::SUPPORTED_EXTENSIONS;
+        use itertools::Itertools;
+        use smoltcp::wire::IpProtocol;
+        for n_extensions in 0..SUPPORTED_EXTENSIONS.len() {
+            for extensions in
+                SUPPORTED_EXTENSIONS.into_iter().permutations(n_extensions)
+            {
+                // Generate a full IPv6 test packet, but pull out the extension
+                // headers as a byte array.
+                let (buf, ipv6_header_size) =
+                    generate_test_packet(extensions.as_slice());
+                let next_hdr =
+                    *(extensions.first().unwrap_or(&IpProtocol::Tcp));
+                let extension_headers = &buf[IPV6_HDR_SZ..ipv6_header_size];
+
+                // Append a TCP header
+                let body = vec![];
+                let mut tcp = TcpHdr::new(3839, 80);
+                tcp.set_seq(4224936861);
+                let ip6 = Ipv6Hdr::new_tcp(
+                    &tcp,
+                    &body,
+                    next_hdr,
+                    &extension_headers,
+                    SRC_IP6,
+                    DST_IP6,
+                );
+                let eth = EtherHdr::new(EtherType::Ipv6, SRC_MAC, DST_MAC);
+
+                let mut packet_buf = Vec::with_capacity(1024);
+                packet_buf.extend_from_slice(&eth.as_bytes());
+                packet_buf.extend_from_slice(&ip6.as_bytes());
+                packet_buf.extend_from_slice(&tcp.as_bytes());
+
+                let generated_packet = Packet::copy(&packet_buf);
+                generated_packet.parse().unwrap();
```

Comment:
You'll want to assert header offsets here. That's why you didn't catch the bug in `parse_hg_ipv6()` when it was still using fixed size. For example, I was able to induce errors by changing that method back to the the previous version and adding the following assertions:

```Rust
                let parsed = generated_packet.parse().unwrap();

                let offsets = &parsed.state.hdr_offsets;
                assert_eq!(offsets.inner.ip.as_ref().unwrap().seg_idx, 0);
                assert_eq!(
                    offsets.inner.ip.as_ref().unwrap().seg_pos,
                    ETHER_HDR_SZ
                );
                assert_eq!(
                    offsets.inner.ip.as_ref().unwrap().pkt_pos,
                    ETHER_HDR_SZ
                );
                assert_eq!(offsets.inner.ulp.as_ref().unwrap().seg_idx, 0);
                assert_eq!(
                    offsets.inner.ulp.as_ref().unwrap().seg_pos,
                    ETHER_HDR_SZ + ipv6_header_size
                );
                assert_eq!(
                    offsets.inner.ulp.as_ref().unwrap().pkt_pos,
                    ETHER_HDR_SZ + ipv6_header_size
                );
```

That's just my example, feel free to code that up differently if you think there is a cleaner way. Also probably should add check for the ether header offset just to be safe.

Reply:
Thanks for taking a look. Makes sense, I think I just didn't see the other tests that make similar assertions, and left it out of this one. I'll add those in now.